### PR TITLE
trim snippets with respect to indentation

### PR DESCRIPTION
**Pull request by yonip23**
See the description on pr 1805 in our binary project for more details.

### CODE REVIEW

File: src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    }
+}
+
+/**
+ * Finds the amount of spaces or tabs in the last line of the given `value`,
+ * returning `null` if `value` has no newlines, or the last line is not whitespaces only.
+ */
+fun lastLineIndentation(value: String): Int? {
+    try {
+        val lastLine = value.lines().last()
```
Comment (boaz-codota) [inline src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt:22]: Is there a way to scan from the end for the 1st linebreak? Won't it be faster for larger texts?

File: src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        if (lastLine.isBlank()) {
+            return lastLine.length
+        }
+    } catch (e: java.util.NoSuchElementException) {
```
Comment (boaz-codota) [inline src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt:26]: When is this exception raised? Upon no lines at all?

File: src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
+package com.tabnine.binary.requests.autocomplete
+
+fun postprocess(request: AutocompleteRequest, result: AutocompleteResponse, tabSize: Int) {
```
Comment (boaz-codota) [inline src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt:3]: I think `tabSize` is constant, and we should calculate the entire regex only once. As far as I know, compiling regexes could have a performance penalty.

Comment (boaz-codota) [reply src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt:3]: And it is best if we can calculate it on plugin start.

File: src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    result.results.forEach { it.new_prefix = it.new_prefix.replace("\t", " ".repeat(tabSize)) }
+
+    val requestIndentation = lastLineIndentation(request.before.replace("\t", " ".repeat(tabSize))) ?: return
+    if (requestIndentation == 0) return
+
+    val regex = constructRegex(requestIndentation)
+    result.results.forEach { entry ->
+        val calculateTrimmingIndex = calculateTrimmingIndex(entry.new_prefix, regex)
+        calculateTrimmingIndex?.let { entry.new_prefix = entry.new_prefix.take(it) }
```
Comment (boaz-codota) [inline src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt:12]:
```suggestion
        calculateTrimmingIndex(entry.new_prefix, regex)?.let { trimmingIndex -> entry.new_prefix = entry.new_prefix.take(trimmingIndex) }
```

File: src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt [hunk 5]
```diff
@@ -0,0 +1,48 @@
+package com.tabnine.binary.requests.autocomplete
+
+fun postprocess(request: AutocompleteRequest, result: AutocompleteResponse, tabSize: Int) {
+    result.results.forEach { it.new_prefix = it.new_prefix.replace("\t", " ".repeat(tabSize)) }
+
+    val requestIndentation = lastLineIndentation(request.before.replace("\t", " ".repeat(tabSize))) ?: return
+    if (requestIndentation == 0) return
```
Comment (boaz-codota) [inline src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt:7]: If you have something like that prefix:
```kotlin
/**
 * Finds the amount of spaces or tabs in the last line of the given `value`,
 * returning `null` if `value` has no newlines, or the last line is not whitespaces only.
 */
fun lastLineIndentation(value: String): Int? {
    try {
        val lastLine = value.lines().last()
```

Where would the following snippet end?
```kotlin
        if (lastLine.isBlank()) {
            return lastLine.length
        }
    } catch (e: java.util.NoSuchElementException) {
        // do nothing
    }
    return null
}

/**
 * Constructs a regex which accepts a \n followed by at most `indentation - 1` spaces,
 * followed by any text or another \n.
 */
fun constructRegex(indentation: Int): Regex {
    val upperLimit = indentation - 1
    return Regex("""^ {0,$upperLimit}[\w\n]+""", RegexOption.MULTILINE)
}
```

File: src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+fun constructRegex(indentation: Int): Regex {
+    val upperLimit = indentation - 1
+    return Regex("""^ {0,$upperLimit}[\w\n]+""", RegexOption.MULTILINE)
+}
+
+/**
+ * Finds the first match of the given `regex` in `result`, *after* the first newline appearance
+ */
+fun calculateTrimmingIndex(result: String, regex: Regex): Int? {
+    val indexOfFirstNewline = result.indexOf("\n")
```
Comment (boaz-codota) [inline src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt:45]: Send a char and not a string
```suggestion
    val indexOfFirstNewline = result.indexOf('\n')
```

File: src/main/java/com/tabnine/prediction/CompletionFacade.java [hunk 7]
```diff
@@ -74,7 +78,13 @@ private AutocompleteResponse retrieveCompletions(
     req.line = document.getLineNumber(offset);
     req.character = offset - document.getLineStartOffset(req.line);
 
-    return binaryRequestFacade.executeRequest(req, determineTimeoutBy(req.before));
+    AutocompleteResponse autocompleteResponse =
+        binaryRequestFacade.executeRequest(req, determineTimeoutBy(req.before));
+
+    if (autocompleteResponse != null) {
+      postprocess(req, autocompleteResponse, tabSize(editor));
+    }
```
Comment (boaz-codota) [inline src/main/java/com/tabnine/prediction/CompletionFacade.java:86]:
```suggestion
    autocompleteResponse?.let { postprocess(req, it, tabSize(editor)); };
```

File: src/test/java/com/tabnine/plugin/completionPostProcess/Spaces.kt [hunk 8]
```diff
@@ -0,0 +1,51 @@
+package com.tabnine.plugin.completionPostProcess
+
+import com.tabnine.binary.requests.autocomplete.postprocess
+import org.junit.Test
+
+class Spaces {
```
Comment (boaz-codota) [inline src/test/java/com/tabnine/plugin/completionPostProcess/Spaces.kt:6]: kapara give it a bit of a more meaningful name

**Review by boaz-codota — APPROVED** [review]
LGTM, but I think we ought to create some performance tests / benchmarks for it. Suggestions are mission critical, and I think we need to act more carefully here. This is another bonus point for Rust as well...

File: src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt (hunk 3 above)
Comment (yonip23) [reply src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt:5]: no, depends on the project / file you're working on - can also be changed by the user..

File: src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt (hunk 5 above)
Comment (yonip23) [reply src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt:14]: its not clear where exactly is the cursor in your example, but assuming you're here:
```kotlin
/**
 * Finds the amount of spaces or tabs in the last line of the given `value`,
 * returning `null` if `value` has no newlines, or the last line is not whitespaces only.
 */
fun lastLineIndentation(value: String): Int? {
    try {
        val lastLine = value.lines().last()
        //<-- cursor here
```

then the snippet will end after the closing bracket of he `if` statement - before the `catch` statement.

File: src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt (hunk 2 above)
Comment (yonip23) [reply src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt:33]: image

Comment (yonip23) [reply src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt:33]: this is so stupid to throw an exception if the list is empty... why doesn't kotlin return `T?` 😢

File: src/main/java/com/tabnine/prediction/CompletionFacade.java (hunk 7 above)
Comment (yonip23) [reply src/main/java/com/tabnine/prediction/CompletionFacade.java:86]: dis is java here 🤮

File: src/test/java/com/tabnine/plugin/completionPostProcess/Spaces.kt (hunk 8 above)
Comment (yonip23) [reply src/test/java/com/tabnine/plugin/completionPostProcess/Spaces.kt:6]: ah, the test's full name is combined with the package name - which gives it the meaning

File: src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt (hunk 1 above)
Comment (yonip23) [reply src/main/java/com/tabnine/binary/requests/autocomplete/CompletionPostprocess.kt]: good point - can use somthing like:

```kotlin
value.substring(value.lastIndexOf("\n"))
```

**Comment by yonip23** [comment]
> LGTM, but I think we ought to create some performance tests / benchmarks for it. Suggestions are mission critical, and I think we need to act more carefully here. This is another bonus point for Rust as well...

I get your concern, but I made sure this code will only run from `Snippet` completions - in which the extra few ms are not critical what so ever

**Pull request merged**